### PR TITLE
Created EngineExtension with default implementation from EnginePlugin

### LINKED ISSUE #323: [FEATURE]Create EngineExtension interface
**Is your feature request related to a problem?**

Extension points (see https://github.com/opensearch-project/opensearch-sdk-java/issues/315) will be (optionally) implemented on Extensions. Eventually these points may need to be communicated to OpenSearch to be registered in the appropriate service or module.

Allowing extensions to inherit from appropriate interfaces depending on their type will match the current Plugin implementation.

**What solution would you like?**

Add a new interface EngineExtension with default implementations for extension points matching EnginePlugin.java.

### PR DESCRIPTION
**Pull request by varuntumbe**
Signed-off-by: varuntumbe <varuntumbe1@gmail.com>

### Description
Add a new interface EngineExtension with default implementations for extension points matching EnginePlugin.java.

### Issues Resolved
#323 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/EngineExtension.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+     * EXPERT:
+     * When an index is created this method is invoked for each engine Extension. Engine Extensions can inspect the index settings
+     * to determine if a custom {@link CodecService} should be provided for the given index. A Extension that is not overriding
+     * the {@link CodecService} through the Extension can ignore this method and the Codec specified in the {@link IndexSettings}
+     * will be used.
+     *
+     * @deprecated Please use {@code getCustomCodecServiceFactory()} instead as it provides more context for {@link CodecService}
+     * instance construction.
+     */
+    @Deprecated
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/EngineExtension.java:51]: Let's not carry over the deprecated API point.

**Review by dbwiddis — COMMENTED** [review]
LGTM except this is a good time to delete that deprecated extension point.

This is also a great example of an "extension-side" implementation that is one of the main reasons for #319.

**Review by dbwiddis — APPROVED** [review]
LGTM! Thanks again!

**Comment by dbwiddis** [comment]
Er, spoke too soon, looks like spotless is complaining about something. :(

**Comment by varuntumbe** [comment]
Yep I think its best to run the spotlessApply before every commit

**Review by dbwiddis — APPROVED** [review]

**Comment by dbwiddis** [comment]
> Yep I think its best to run the spotlessApply before every commit

I have a single key macro for "gradle spotlessapply check". :)

**Comment by varuntumbe** [comment]
Will take some time for #319 , I need to read and understand ExtensionManager class and will ask questions If I get stuck understanding the code

**Review by saratvemulapalli — APPROVED** [review]

**Pull request merged**